From ac995c581e0e493bc4fc4fc69e7a0e0f6b674577 Mon Sep 17 00:00:00 2001
From: Deepak Rathore <deepak.rathore@einfochips.com>
Date: Thu, 19 Jan 2023 15:59:07 +0530
Subject: [PATCH 1/2] Added AR0830 camera sensor support

- Added AR0830 Camera sensor support
- Added AR0830 sensor details in ap1302 driver

Change-Id: I0db658b926e9a4626bbd0e81f2de3c20ccd3428a
Signed-off-by: Deepak Rathore <deepak.rathore@einfochips.com>
---
 arch/arm/boot/dts/Makefile                    |  1 +
 ...mp15xx-avenger96-overlay-ap1302-ar0830.dts | 95 +++++++++++++++++++
 drivers/media/i2c/ap1302.c                    | 17 +++-
 3 files changed, 112 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar0830.dts

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 85311506d..afaae85f1 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1077,6 +1077,7 @@ dtb-$(CONFIG_ARCH_STM32) += \
 	stm32mp15xx-avenger96-overlay-ap1302-ar0430.dtbo \
 	stm32mp15xx-avenger96-overlay-ap1302-arx3a0.dtbo \
 	stm32mp15xx-avenger96-overlay-ap1302-ar1335.dtbo \
+	stm32mp15xx-avenger96-overlay-ap1302-ar0830.dtbo \
 	stm32mp157a-dk1.dtb \
 	stm32mp157a-iot-box.dtb \
 	stm32mp157a-stinger96.dtb \
diff --git a/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar0830.dts b/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar0830.dts
new file mode 100644
index 000000000..61358d98f
--- /dev/null
+++ b/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar0830.dts
@@ -0,0 +1,95 @@
+#include <dt-bindings/clock/stm32mp1-clks.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/dts-v1/;
+/plugin/;
+
+&{/} {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	sram@10050000 {
+               compatible = "mmio-sram";
+               reg = <0x10050000 0x10000>;
+               #address-cells = <1>;
+               #size-cells = <1>;
+               ranges = <0 0x10050000 0x10000>;
+
+               dma_pool: dma_pool@0 {
+                       reg = <0x0 0x10000>;
+                       pool;
+               };
+         };
+};
+
+&dma1 {
+	sram = <&dma_pool>;
+};
+
+&dma2 {
+	sram = <&dma_pool>;
+};
+
+&adc {
+	status = "disabled";
+};
+
+&adc1{
+	status = "disabled";
+};
+
+&adc2{
+	status = "disabled";
+};
+
+&i2c2 {
+	ap1302: camera@3d {
+            compatible = "on,ap1302";
+            reg = <0x3d>;
+            pinctrl-names = "default", "sleep";
+            pinctrl-0 = <&rcc_pins_a>;
+            pinctrl-1 = <&rcc_sleep_pins_a>;
+            clocks = <&rcc CK_MCO1>;
+            clock-names = "xclk";
+            assigned-clocks = <&rcc CK_MCO1>;
+            assigned-clock-rates = <24000000>;
+            reset-gpios = <&gpioc 3 GPIO_ACTIVE_LOW>;
+            standby-gpios = <&gpiod 14 GPIO_ACTIVE_HIGH>;
+            i2csel-gpios = <&gpiof 12 GPIO_ACTIVE_HIGH>;
+            isptrig-gpios = <&gpioa 12 GPIO_ACTIVE_HIGH>;
+            status = "okay";
+
+	    port {
+                    ap1302_0: endpoint {
+                            remote-endpoint = <&stmipi_0>;
+                            data-lanes = <1 2>;
+                    };
+            };
+
+	    sensors {
+                #address-cells = <1>;
+                #size-cells = <0>;
+		onnn,model = "onnn,ar0830";
+                    sensor@0 {
+                        reg = <0>;
+                   };
+            };
+      };
+};
+
+&stmipi{
+	status = "okay";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			stmipi_0: endpoint {
+				data-lanes = <1 2>;
+				remote-endpoint = <&ap1302_0>;
+			};
+		};
+	};
+};
diff --git a/drivers/media/i2c/ap1302.c b/drivers/media/i2c/ap1302.c
index 6e2fb087d..8ebb572cb 100644
--- a/drivers/media/i2c/ap1302.c
+++ b/drivers/media/i2c/ap1302.c
@@ -604,7 +604,22 @@ static const struct ap1302_sensor_info ap1302_sensor_info[] = {
 			{ "vddio", 0 },
 			{ NULL, 0 },
 		},
-	}
+	}, {
+		.model = "onnn,ar0830",
+		.name = "ar0830",
+		.i2c_addr = 0x36,
+		.resolution = { 3840, 2160},
+		.fps = 5,
+		.format = MEDIA_BUS_FMT_RGB565_2X8_LE,
+		.supplies = (const struct ap1302_sensor_supply[]) {
+                        { "vddpll", 0 },
+                        { "vaa", 0 },
+                        { "vdd", 0 },
+                        { "vddio", 0 },
+                        { NULL, 0 },
+		},
+	},
+
 };
 
 static const struct ap1302_sensor_info ap1302_sensor_info_tpg = {
-- 
2.17.1

